pkg/scheduler/internal/heap/heap.go

```diff
@@ -114,56 +113,50 @@ func (h *data) Pop() interface{} {
 	return item.obj
 }
 
-// Peek is supposed to be called by heap.Peek only.
-func (h *data) Peek() interface{} {
+// Peek returns the head of the heap without removing it.
+func (h *data[T]) Peek() T {
 	if len(h.queue) > 0 {
 		return h.items[h.queue[0]].obj
 	}
-	return nil
+	var zero T
+	return zero
 }
 
 // Heap is a producer/consumer queue that implements a heap data structure.
 // It can be used to implement priority queues and similar data structures.
-type Heap struct {
+type Heap[T any] struct {
 	// data stores objects and has a queue that keeps their ordering according
 	// to the heap invariant.
-	data *data
+	data *data[T]
 	// metricRecorder updates the counter when elements of a heap get added or
 	// removed, and it does nothing if it's nil
 	metricRecorder metrics.MetricRecorder
 }
 
 // Add inserts an item, and puts it in the queue. The item is updated if it
 // already exists.
-func (h *Heap) Add(obj interface{}) error {
-	key, err := h.data.keyFunc(obj)
-	if err != nil {
-		return cache.KeyError{Obj: obj, Err: err}
-	}
+func (h *Heap[T]) Add(obj T) {
+	key := h.data.keyFunc(obj)
 	if _, exists := h.data.items[key]; exists {
 		h.data.items[key].obj = obj
 		heap.Fix(h.data, h.data.items[key].index)
 	} else {
-		heap.Push(h.data, &itemKeyValue{key, obj})
+		heap.Push(h.data, &itemKeyValue[T]{key, obj})
 		if h.metricRecorder != nil {
 			h.metricRecorder.Inc()
 		}
 	}
-	return nil
 }
 
 // Update is the same as Add in this implementation. When the item does not
 // exist, it is added.
```

Comment:
By the way, what's the reason behind we have both Add and Update while they're the same? Can we just have one of them?

Reply:
I would keep it for verbosity, to prevent making some kind of `Delete`-`Add` combo accidentally in the future.

Reply:
If your concern is that the naming would be confusing if we only have `Add` (or `Update`), can we name the function `Set`, or `AddOrUpdate` if we go more straightforward?

Reply:
`AddOrUpdate` makes sense, renamed